query/regexp.go

```diff
@@ -42,3 +42,60 @@ func LowerRegexp(r *syntax.Regexp) *syntax.Regexp {
 
 	return &newRE
 }
+
+// optimizeRegexp converts capturing groups to non-capturing groups.
+// Returns original input if an error is encountered
+func optimizeRegexp(re *syntax.Regexp) *syntax.Regexp {
+	r := convertCapture(re)
+	return r.Simplify()
```

Comment:
can't we avoid this Simplify call if we didn't do anything to r? I'd expect where you return in `convertCapture` to be `return re.Simplify()`

Reply:
For sure, that call to Simplify could be avoided.
As there might be more optimisations in the future I put the Simplify at the end of `optimizeRegexp`.
Do you think a Simplify at the end of each potential optimisation would be a more preferable pattern?

Reply:
> Do you think a Simplify at the end of each potential optimisation would be a more preferable pattern

maybe, but we can cross that bridge when we actually add more optimizations :) I suppose a simplify each time will make it clearer to understand what each stage of optimizations is working with.